src/FESpaces/ConformingFESpaces.jl

```diff
@@ -189,10 +199,8 @@ end
 
 function _compute_cell_space(
```

Comment:
Is this method required after the refactoring? I would remove it since it only works for lagrangian FEs and the method `_compute_cell_space` defined for `CellFE` should be enough.

Reply:
That is

- Keep this: ` _compute_cell_space(cell_fe::CellFE)` (removing trian)
- Delete this: ` _compute_cell_space(cell_reffe::AbstractArray{<:ReferenceFE}, trian::Triangulation)
`

Reply:
> Is this method required after the refactoring? 

I think so, it is called at several places. The method returns a pair with the `fe_basis` and the `dof_basis` (CellDatum objects). For conforming FE spaces, these objects are extracted out of `CellFE`. For other FE spaces (I would say made invariably of Lagrangian FEs), these objects are built out of an array of RefFEs, and the triangulation. 

Thus: (1) the method cannot be removed. (2) the trian parameter cannot be removed. (3) the method does not only work with lagrangian FEs.  

On the other hand, the `trian` parameter can be removed for both variants of `_compute_cell_space` (I would keep the second variant, it is still helpful).

Reply:
There is something that I don't understand. After the refactoring, you need a model to get the cell-wise shape functions from the reference FEs in general. Thus, you cannot exteact valid cell-wise shape functions from the arguments in this method `_compute_cell_space(cell_reffe::AbstractArray{<:ReferenceFE}, trian::Triangulation)` in the general case... the model is missing.

Reply:
is perhaps being used in discontinus spaces? Even in discontinus spaces, I think you need to apply the piola transformation (perhaps the sign flip is not hat crucial in this case). Thus, you still need the model.

We need to remove the method _compute_cell_space(cell_reffe::AbstractArray{<:ReferenceFE}, trian::Triangulation) since i believe it does not make sense anymore after the refactoring. Just go to all the calles of it (or the calles of compute_cell_space) and transform the `AbstractArray{<:ReferenceFE}` into a `CellFE` using a model (which should be over there). Then the method could be removed.

Reply:
> There is something that I don't understand. After the refactoring, you need a model to get the cell-wise shape functions from the reference FEs in general. Thus, you cannot extract valid cell-wise shape functions from the arguments in this method _compute_cell_space(cell_reffe::AbstractArray{<:ReferenceFE}, trian::Triangulation) in the general case... the model is missing.

`compute_cell_space` (that in turn calls `_compute_cell_space`) is also called from FESpaces which are not created out of a model (e.g., DG FE Spaces), but out of a triangulation instead. In this case, compute_cell_space gets the array of RefFEs and the triangulation. (i.e., you cannot pass the model, because it is not available).

Reply:
In the case of FESpaces which are created out of a model, `compute_cell_space` gets a `CellFE` instance (instead of an array of reffes, that in turn has inside as member variables the shape functions and the dof basis created out of a model.

Reply:
It looks like you can only create a dg space from the triangulation for Lagranges FEs, not for Raviart-Thomas or in general.

After the refactoring, the construction of the cell-wise shape functions and dofs basis is coupled with the discrete model, thus it is not possible to build a dg space from a triangulaiton in general now....

Another reason why i don't like this new coupling, but It seems that we need to deal with it.

Reply:
>  is also called from FESpaces which are not created out of a model (e.g., DG FE Spaces)

It looks like we need to remove the constructor of the DG space from a `Triangulation`, and replace it from 2 new constructors, one that takes a `DiscreteModel`, and another that takes the a `CellFE`. The former will call the latter.

Reply:
> We need to remove the method _compute_cell_space(cell_reffe::AbstractArray{<:ReferenceFE}, trian::Triangulation) since i believe it does not make sense anymore after the refactoring. Just go to all the callers of it (or the calles of compute_cell_space) and transform the AbstractArray{<:ReferenceFE} into a CellFE using a model (which should be over there). Then the method could be removed.

I guess that this comment no longer applies, right?

Reply:
> You can only create a dg space from the triangulation for Lagranges FEs, not for Raviart-Thomas or in general.

> After the refactoring, the construction of the cell-wise shape functions and dofs basis is coupled with the discrete model, thus it is not possible to build a dg space from a triangulaiton in general now....

Could this be addressed defining variants of `get_cell_shapefuns` and `get_cell_dof_basis` for an array of reffes + triangulation?

Reply:
It still applies. the DG space cannot constructed from a triangulation anymore.

We need to remove the constructor of the DG space from a Triangulation, and replace it from 2 new constructors, one that takes a DiscreteModel, and another that takes the arrays cell_shapefuns, cell_dof_basis. The former will call the latter.

In addition, remove this 
`_compute_cell_space(cell_reffe::AbstractArray{<:ReferenceFE}, trian::Triangulation)`
and replace it by 
`_compute_cell_space(cell_shapefuns,cell_dof_basis,trian)` (trian only used to build the CellDatum objects)

The same pattern should be used in other FEspaces that are build from a triangulation (except the spaces that are tailored for Lagrangian elements)

Reply:
> Could this be addressed defining variants of get_cell_shapefuns and get_cell_dof_basis for an array of reffes + triangulation?

I would not do this for the same reason now we only have get_shapefuns(reffe) and not `get_shapefuns(reffe,map)`

Reply:
> It looks like we need to remove the constructor of the DG space from a Triangulation, and replace it from 2 new constructors, one that takes a DiscreteModel, and another that takes the arrays cell_shapefuns, cell_dof_basis. The former will call the latter.

For my understanding. Why the previous code was trying to stick into FE spaces built out of Triangulations as far as possible? What do you gain with that? (I am sure there is something, but I do not see it now). I guess that you gain generality/flexibility ... but do you have use cases in mind where this is necessary/relevant?

Reply:
> Why the previous code was trying to stick into FE spaces built out of Triangulations as far as possible?

Because a triangulation is cheaper tho build than a discrete model. But it practice we always build a discrete model, so not bit deal...

I am even considering to add a discrete model into CellDatum objects instead of a triangulation (it would simplify the implementation of mixed dimensional pdes, and perhaps hybrid methods also)

Reply:
Ok. So recap after this long discussion, it seems we have came up with two options:

1. Discontinuous FE spaces (except those built out of Lagrangian RefFEs) must be built out of a model.
  * Pros: we stick into a single interface for cell_shapefuns, cell_dof_basis.
  * Cons: forced to use a model (do we loose something?). We end up doing work which is not needed, e.g., sign flip. 
2. Define variants of get_cell_shapefuns and get_cell_dof_basis for an array of reffes + triangulation.
* Pros: we preserve the invariant of being able to set up the FE Space out of triangulation. We can skip doing work which is not needed for DG spaces (e.g., change of basis for Lagrangian FEs, sign flip for RT FEs). 
* Cons: "I would not do this for the same reason now we only have get_shapefuns(reffe) and not get_shapefuns(reffe,map)" the dispatching among Discontinuous and Continuous FE Spaces is based on the type of geometry object (i.e., model versus triangulation). This seems to me a bit ugly.  

Wrote for documentation purposes. If you feel that 1. is the way to go, we can proceed. Perhaps we can think a little bit more. I am going to stop working now on this. (it's late already :-))